ruma-signatures/src/functions.rs

```diff
@@ -731,13 +744,75 @@ pub fn redact(
     Ok(event)
 }
 
+/// Extracts the server names to check signatures for given event. It will return the sender's
+/// server (unless it's a third party invite) and the event id server (on v1 and v2 room versions)
+fn servers_to_check_signatures(
+    object: &CanonicalJsonObject,
+    version: &RoomVersionId,
+) -> Result<HashSet<String>, Error> {
+    let mut servers_to_check = HashSet::new();
+
+    if !is_third_party_invite(object)? {
+        match object.get("sender") {
+            Some(CanonicalJsonValue::String(raw_sender)) => {
+                let user_id = UserId::from_str(raw_sender)
+                    .map_err(|_| Error::new("could not parse user id"))?;
+
+                servers_to_check.insert(user_id.server_name().to_string());
+            }
+            _ => return Err(Error::new("field `sender` must be a JSON string")),
+        };
+    }
+
+    match version {
+        RoomVersionId::Version1 | RoomVersionId::Version2 => match object.get("event_id") {
+            Some(CanonicalJsonValue::String(raw_event_id)) => {
+                let event_id = EventId::from_str(raw_event_id)
+                    .map_err(|_| Error::new("could not parse event id"))?;
+
+                let server_name = event_id
+                    .server_name()
+                    .ok_or_else(|| {
+                        Error::new("Event id should have a server name for the given room version")
+                    })?
+                    .to_string();
+
+                servers_to_check.insert(server_name);
+            }
+            _ => {
+                return Err(Error::new(
+                    "Expected to find a string `event_id` for the given room version",
+                ))
+            }
+        },
+        _ => (),
+    }
+
+    Ok(servers_to_check)
+}
+
+/// Checks if `object` contains an event of type `m.room.third_party_invite`
+fn is_third_party_invite(object: &CanonicalJsonObject) -> Result<bool, Error> {
+    match object.get("type") {
+        Some(CanonicalJsonValue::String(raw_type)) => Ok(raw_type == "m.room.third_party_invite"),
```

Comment:
I had some doubts about adding ruma-events as a dependency to avoid hard-coding this event type. Is there any issue with adding it?

Reply:
On one hand, it seems like overkill to include ruma-events for that; on the other hand I doubt ruma-signatures is used without ruma-events by anyone (the only benefit decoupling would still have then is build parallelization).

I don't much care either way.

Reply:
As you don't have any strong opinion, I prefer to avoid adding the dependency and leave it as it is now.